tool attribute 1.1

shedaniel opened the pull request.

LemmaEOF reviewed (approved):
Looks good; no breaking changes (just a verbump for the cancellable issue and a fixed typo in a mixin method name)

liach commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/mixin/tool/attribute/MixinPlayerInventory.java:
```diff
@@ -63,7 +63,7 @@ public void actMiningLevel(BlockState state, CallbackInfoReturnable<Boolean> inf
 
 	@Inject(method = "getBlockBreakingSpeed", at = @At("HEAD"), cancellable = true)
-	public void actMiningSleed(BlockState state, CallbackInfoReturnable<Float> info) {
+	public void actMiningSpeed(BlockState state, CallbackInfoReturnable<Float> info) {
```
injector can be private

shedaniel committed:
Actual support

shedaniel committed:
Remove this useless mixin

shedaniel commented:
Major Changes:
- Fixes modded tools on vanilla blocks by hardcoding mining levels for vanilla blocks
- Partial support for shears (working mining level and break speed, broken leaves drops as they check shears itself in datapacks)
- DynamicAttributeTool methods now accept BlockState, useful for different speed for different blocks (shears does this)
- Added `getMiningLevel` in `ToolManager.Entry`, this still **shouldn't** be used because this is still in the impl package, I plan to have an API facing part of this in the future.
- Bump version to `2.0.0`

Juuxel commented:
Wait, this isn't the PR I approved... 🤔

LemmaEOF commented on fabric-tool-attribute-api-v1/build.gradle:
```diff
@@ -1,7 +1,8 @@
 archivesBaseName = "fabric-tool-attribute-api-v1"
-version = getSubprojectVersion(project, "1.0.1")
+version = getSubprojectVersion(project, "2.0.0")
```
I'm never getting another PR merged again, am I :tiny_potato:

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/VanillaToolsSupport.java:
```diff
@@ -0,0 +1,63 @@
+		registerSupport(FabricToolTags.PICKAXES, 3, Blocks.OBSIDIAN);
+		registerSupport(FabricToolTags.PICKAXES, 2, Blocks.DIAMOND_BLOCK, Blocks.DIAMOND_ORE, Blocks.EMERALD_ORE, Blocks.EMERALD_BLOCK, Blocks.GOLD_BLOCK, Blocks.GOLD_ORE, Blocks.REDSTONE_ORE);
+		registerSupport(FabricToolTags.PICKAXES, 1, Blocks.IRON_BLOCK, Blocks.IRON_ORE, Blocks.LAPIS_BLOCK, Blocks.LAPIS_ORE);
+		registerSupport(FabricToolTags.PICKAXES, 0, Blocks.ACTIVATOR_RAIL, Blocks.COAL_ORE, Blocks.COBBLESTONE, Blocks.DETECTOR_RAIL, Blocks.DIAMOND_BLOCK, Blocks.DIAMOND_ORE, Blocks.POWERED_RAIL, Blocks.GOLD_BLOCK, Blocks.GOLD_ORE, Blocks.ICE, Blocks.IRON_BLOCK, Blocks.IRON_ORE, Blocks.LAPIS_BLOCK, Blocks.LAPIS_ORE, Blocks.MOSSY_COBBLESTONE, Blocks.NETHERRACK, Blocks.PACKED_ICE, Blocks.BLUE_ICE, Block ...
```
is this really necessary if we have the registerBlockSupport call above? The other pick mining levels are definitely necessary, but this one seems redundant.

shedaniel replied:
Things like `ACTIVATOR_RAIL` is not in any of the materials affected by pickaxe, I will have to look into every one of these to check if they match the material, might do though.

LemmaEOF replied:
The only ones that need to stay, from what I can tell, are the rails (normal, powered, detector, activator), the ices (normal, packed, and blue), and the shulker boxes (normal plus all 16 colors). All the others use `Material.STONE` or `Material.METAL`.

shedaniel replied:
Done, please check

LemmaEOF replied:
Looks good!

shedaniel commented:
also the current latest fabric version on CF is broken, it turns your tool into a multi-tool, grab a diamond pickaxe and you can mine grass, this PR fixes that as well

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/VanillaToolsSupport.java:
```diff
@@ -0,0 +1,63 @@
+			registerBlockSupport(Registry.BLOCK.getId(block), block);
+		}
+
+		RegistryEntryAddedCallback.event(Registry.BLOCK).register((rawId, id, object) -> registerBlockSupport(id, object));
```
I'm not sure where in loading stages ModInitializer gets called - is it before vanilla reg, after it, or unreliable?

Juuxel replied:
Mod initializers are run after vanilla registries have been bootstrapped.

shedaniel replied:
Ah right, I will remove this

LemmaEOF reviewed (changes requested):
Looking good, but the vanilla support probably doesn't need both auto-populate *and* hardcoding every single value. Hardcoding should only be necessary if vanilla hardcodes those same specific blocks, like for pick mining levels.

immibis commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -48,7 +69,28 @@
+			Block block = state.getBlock();
+			if (block == Blocks.COBWEB || state.matches(BlockTags.LEAVES)) return 15f;
+			if (state.matches(BlockTags.WOOL)) return 5f;
+		}
```
Why not put this code in the ShearsItem mixin?

shedaniel replied:
Done.

shedaniel committed:
Resolve some issues

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -38,7 +40,23 @@
 	 */
 	//TODO: nullable on user once we have an official @Nullable annotation in
-	int getMiningLevel(ItemStack stack, LivingEntity user);
+	@Deprecated
```
Add javadocs for the deprecation, maybe?

shedaniel committed:
Remove duplicate entries

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/VanillaToolsSupport.java:
```diff
@@ -0,0 +1,61 @@
+		registerSupport(FabricToolTags.PICKAXES, 1, Blocks.IRON_BLOCK, Blocks.IRON_ORE, Blocks.LAPIS_BLOCK, Blocks.LAPIS_ORE);
+		registerSupport(FabricToolTags.PICKAXES, 0, Blocks.ACTIVATOR_RAIL, Blocks.COAL_ORE, Blocks.COBBLESTONE, Blocks.DETECTOR_RAIL, Blocks.DIAMOND_BLOCK, Blocks.DIAMOND_ORE, Blocks.POWERED_RAIL, Blocks.GOLD_BLOCK, Blocks.GOLD_ORE, Blocks.ICE, Blocks.IRON_BLOCK, Blocks.IRON_ORE, Blocks.LAPIS_BLOCK, Blocks.LAPIS_ORE, Blocks.MOSSY_COBBLESTONE, Blocks.NETHERRACK, Blocks.PACKED_ICE, Blocks.BLUE_ICE, Block ...
+
+		registerSupport(FabricToolTags.AXES, 0, Blocks.OAK_PLANKS, Blocks.SPRUCE_PLANKS, Blocks.BIRCH_PLANKS, Blocks.JUNGLE_PLANKS, Blocks.ACACIA_PLANKS, Blocks.DARK_OAK_PLANKS, Blocks.BOOKSHELF, Blocks.OAK_WOOD, Blocks.SPRUCE_WOOD, Blocks.BIRCH_WOOD, Blocks.JUNGLE_WOOD, Blocks.ACACIA_WOOD, Blocks.DARK_OAK_WOOD, Blocks.OAK_LOG, Blocks.SPRUCE_LOG, Blocks.BIRCH_LOG, Blocks.JUNGLE_LOG, Blocks.ACACIA_LOG,  ...
```
The ones necessary to keep here are all three types of pumpkin, melons, scaffolding, ladders and all wood buttons.

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/VanillaToolsSupport.java:
```diff
@@ -0,0 +1,61 @@
+
+		registerSupport(FabricToolTags.AXES, 0, Blocks.OAK_PLANKS, Blocks.SPRUCE_PLANKS, Blocks.BIRCH_PLANKS, Blocks.JUNGLE_PLANKS, Blocks.ACACIA_PLANKS, Blocks.DARK_OAK_PLANKS, Blocks.BOOKSHELF, Blocks.OAK_WOOD, Blocks.SPRUCE_WOOD, Blocks.BIRCH_WOOD, Blocks.JUNGLE_WOOD, Blocks.ACACIA_WOOD, Blocks.DARK_OAK_WOOD, Blocks.OAK_LOG, Blocks.SPRUCE_LOG, Blocks.BIRCH_LOG, Blocks.JUNGLE_LOG, Blocks.ACACIA_LOG,  ...
+
+		registerSupport(FabricToolTags.SHOVELS, 0, Blocks.CLAY, Blocks.DIRT, Blocks.COARSE_DIRT, Blocks.PODZOL, Blocks.FARMLAND, Blocks.GRASS_BLOCK, Blocks.GRAVEL, Blocks.MYCELIUM, Blocks.SAND, Blocks.RED_SAND, Blocks.SNOW_BLOCK, Blocks.SNOW, Blocks.SOUL_SAND, Blocks.GRASS_PATH, Blocks.WHITE_CONCRETE_POWDER, Blocks.ORANGE_CONCRETE_POWDER, Blocks.MAGENTA_CONCRETE_POWDER, Blocks.LIGHT_BLUE_CONCRETE_POWDE ...
```
We could probably trim this down a lot by having `Material.EARTH` and `Material.SAND` be supported materials for shovels. That'd cover everything except clay, grass, mycelium, snow and snow blocks. You've also got `Blocks.SNOW` and `Blocks.SNOW_BLOCK` in here twice.

LemmaEOF reviewed (approved):
I've gotten the lists for all the materials that can be auto-registered.

modmuss50 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/VanillaToolsSupport.java:
```diff
@@ -0,0 +1,61 @@
+			registerBlockSupport(Registry.BLOCK.getId(block), block);
+		}
+
+		registerSupport(FabricToolTags.PICKAXES, 3, Blocks.OBSIDIAN);
```
this seems un-maintainable tbh

Player3324 replied:
The list looks rather arbitrary to me, why not use net.minecraft.item.AxeItem.EFFECTIVE_BLOCKS etc?

shedaniel replied:
Actually true, we still need to hardcode the blocks with special mining levels though

LemmaEOF replied:
@modmuss50 that's the thing - Mojang's system of hardcoding block mining levels and effectivenesses is fundamentally unmaintainable. We're doing what we can to work around it but if Mojang hardcodes it we have to too.

shedaniel replied:
We now work around this by using tool handlers, check the latest code and you will see that this is basically completely different

modmuss50 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -38,7 +40,23 @@
+	 * @param user The current user of the tool, or null if there isn't any.
+	 * @return The mining level of the item. 3 is equal to a diamond pick.
+	 */
+	//TODO: nullable on user once we have an official @Nullable annotation in
```
TODO isnt really right for this. just a comment.

shedaniel committed:
maybe?

shedaniel committed:
fix max breaking speed

shedaniel committed:
choose an item that is the closest

shedaniel committed:
Set version to 1.1

shedaniel committed:
better docs remove stuff

shedaniel committed:
Merge remote-tracking branch 'fabric/1.15' into patch-1.15

# Conflicts:
#	fabric-tool-attribute-api-v1/build.gradle
#	fabric-tool-attribute-api-v1/src/main/resources/fabric-tool-attribute-api-v1.mixins.json

shedaniel committed:
deprecation doc and no todo comments

shedaniel committed:
Check if block has valid mining level

shedaniel commented:
Just tested this with modded blocks and modded tools, seems to work fine. Would like to get reviews.

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolManager.java:
```diff
@@ -0,0 +1,43 @@
+import net.fabricmc.fabric.api.util.TriState;
+import net.fabricmc.fabric.impl.tool.attribute.ToolManagerImpl;
+
+public final class ToolManager {
```
Now that these are publicly exposed, they're *definitely* going to need javadocs.

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolManager.java:
```diff
@@ -0,0 +1,43 @@
+	private ToolManager() {
+	}
+
+	public interface ToolHandler {
```
I hope this won't mean that every mod needs to register events for all their tools?

shedaniel replied:
nope, ideally you don't even need to touch this callback, I have thought of keeping this callback internal, but have decided to leave it open in case someone adds a new tool type that is completely separate from vanilla

LemmaEOF replied:
alright, sounds good.

LemmaEOF commented:
@shedaniel anything else this needs to do, and have your tests worked?

shedaniel commented:
@Boundarybreaker Yes, my tests have worked, I do need to write javadocs though.

shedaniel committed:
javadocs

LemmaEOF reviewed (approved):
Looks good to me! I'll toss this over to the triage team; we might want to fast-track this since the current state is broken.

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolManager.java:
```diff
@@ -0,0 +1,96 @@
+		 * @param state the block state to break
+		 * @return the result of effectiveness
+		 */
+		default ActionResult isEffectiveOn(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
```
The parameter order here differs from DynamicAttributeTool

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolManager.java:
```diff
@@ -0,0 +1,96 @@
+		 * @param state the block state to break
+		 * @return the result of mining speed, null if pass.
+		 */
+		default Float getMiningSpeedMultiplier(Tag<Item> tag, ItemStack stack, LivingEntity user, BlockState state) {
```
Same order issue, not sure if Float is a good choice vs TypedActionResult or what it was called?

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolHandlers.java:
```diff
@@ -0,0 +1,343 @@
+		ToolManager.general().register(new VanillaToolsVanillaBlocksToolHandler());
+		ToolManager.tag(FabricToolTags.PICKAXES).register(new ModdedMiningToolsVanillaBlocksToolHandler(
+				ImmutableList.of(
+						createPickaxe(ToolMaterials.WOOD),
```
If my two comments below are viable, we should be able to use plain vanilla item instances directly besides removing quite some code.

shedaniel replied:
I have changed this to use vanilla tools, only changing the material if needed. Don't know if this is a good approach.

Player3324 replied:
This is not feasible since it is changing global state in a multi-threaded environment, causing e.g. integrated server behavior to change while the client thread mines something. Changing material or mining level parameters would first have to be removed.

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolHandlers.java:
```diff
@@ -0,0 +1,343 @@
+				int miningLevel = ((DynamicAttributeTool) stack.getItem()).getMiningLevel(tag, state, stack, user);
+				if (miningLevel < 0) return ActionResult.PASS;
+				ToolItem vanillaItem = getVanillaItem(miningLevel);
+				ToolMaterial tempMaterial = vanillaItem.getMaterial();
```
Messing with the material shouldn't be really necessary anymore, I'd rely on the retrieved vanilla tool as is. Vanilla shouldn't have any blocks that are incompatible with its own tools (=materials)

shedaniel replied:
What if we have tools that have mining levels outside 0 to 3? Or larger than level 1 axes?

Player3324 replied:
The next-best vanilla tool should suffice for vanilla blocks unless I am missing something. Even if the modded tool has level 20, a vanilla diamond tool should still be just as effective considering vanilla won't have level 20 blocks but at worst obsidian.

shedaniel replied:
True, I can do that for vanilla blocks

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolHandlers.java:
```diff
@@ -0,0 +1,343 @@
+				float tempMiningSpeed = ((MiningToolItemAccessor) vanillaItem).getMiningSpeed();
+				((MiningToolItemAccessor) vanillaItem).setMiningSpeed(((DynamicAttributeTool) stack.getItem()).getMiningSpeedMultiplier(tag, state, stack, user));
+				float miningSpeed = vanillaItem.getMiningSpeed(stack, state);
+				((MiningToolItemAccessor) vanillaItem).setMiningSpeed(tempMiningSpeed);
```
Is it really needed to go through the vanilla item again? May a pure effectiveness check as above suffice? If not, can't the return value from the unbiased vanilla tool be analyzed and then adjusted for the desired new speed, potentially with some hardcoded values?

i509VCB replied:
Also some spacing in that method please, it's stuffed to the brim

shedaniel replied:
Don't think we can adjust according to vanilla, for example, swords have constant speed on cobweb no matter the speed of the tool material.

Player3324 replied:
These exceptions may be worth hardcoding if it saves us from all the tool manipulation trickery. Maybe first comparing two vanilla tools (like diamond vs iron) to each other is sufficient to test and observe special behavior.

shedaniel replied:
In some cases, it should be impossible to compare with two vanilla tools, for example obsidian, it is impossible to have two tools to compare with (in 1.15 anyway), for other levels it should be easy.

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,206 @@
+		}
+	}
+
+	private static final Map<Tag<Item>, Event<ToolManager.ToolHandler>> HANDLER_MAP = new HashMap<>();
```
Can this use IdentityHashMap?

shedaniel replied:
Why? Tags with the same identifier are supposed to use the same handler.

Player3324 replied:
I didn't quite remember how we obtained tags, they are not using identity behavior so it can't use IdentityHashMap.

LemmaEOF replied:
Using `IdentityHashMap` might also not be super future-proof, given the big tag changes coming in 1.16.

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,206 @@
+	private static final Map<Block, EntryImpl> entries = new HashMap<>();
+
+	private ToolManagerImpl() {
+	}
```
construction prevention in non-api is superfluous

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,206 @@
+	private static final Map<Tag<Item>, Event<ToolManager.ToolHandler>> HANDLER_MAP = new HashMap<>();
+	private static final Event<ToolManager.ToolHandler> GENERAL_TOOLS_HANDLER = EventFactory.createArrayBacked(ToolManager.ToolHandler.class, ToolManagerImpl::taggedToolHandlerInvoker);
+
+	private static final Map<Block, EntryImpl> entries = new HashMap<>();
```
Also IdentityHashMap

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolManager.java:
```diff
@@ -0,0 +1,96 @@
+	 */
+	public static Event<ToolHandler> general() {
+		return ToolManagerImpl.general();
+	}
```
I don't think those events  are too useful for standard uses, as such they should probably be moved down in this class and the comment should refer to DynamicAttributeTool as the standard way to implement the functionality. The same applies to the class comment.

Player3324 reviewed (commented):
Please explain when/what for the api classes should be used in their comments and include brief example code (oriented towards advanced users/wiki alternative for modders that otherwise know mc+java+basic fabric).

A brief mention wrt how the block side works / what the associated Fabric API is and how it interacts would be another useful addition.

shedaniel committed:
Most of the issues resolved

shedaniel committed:
Some more javadoc and resolve some issues

shedaniel committed:
Forgot to change order for these

shedaniel committed:
Adds post process method to DynamicAttributeTool
Fixed going to vanilla again to query
Fixed tool mining level < 0
Fixed vanilla items on modded blocks not working
Renamed taggedToolHandlerInvoker to toolHandlerInvoker
Renamed entries to ENTRIES

shedaniel committed:
New comparision method for modded tools to vanilla blocks

shedaniel committed:
Remove useless mixins and use getMaterial().getMiningSpeed to support more tools.

shedaniel committed:
Simply mixin

shedaniel committed:
Some useful javadocs

shedaniel commented:
@sfPlayer1 Please review the changes, thanks.

shedaniel committed:
Fix license

shedaniel committed:
Bump fabric-object-builders-v0 because we are good

shedaniel commented:
I would much prefer if #537 is merged so I can rebase my changes on to it

shedaniel committed:
Merge remote-tracking branch 'fabric/1.15' into patch-1

shedaniel committed:
Bump version of fabric-object-builder-api-v1 because we are nice

shedaniel commented:
I've updated this for the new fabric-object-builder-api-v1.

i509VCB commented on fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/block/FabricBlockSettings.java:
```diff
@@ -233,16 +233,27 @@ public FabricBlockSettings collidable(boolean collidable) {
 	}
 
+	/**
+	 * Please make the material require tool if you plan to disable drops and slow the breaking down using the
```
```suggestion
	 * Please make the material requires a tool if you plan to disable drops and slow the breaking down using the
```

LemmaEOF replied:
should be `require a`

i509VCB commented on fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/api/object/builder/v1/block/FabricMaterialBuilder.java:
```diff
@@ -84,6 +84,9 @@ public FabricMaterialBuilder replaceable() {
 	}
 
+	/**
+	 * Make the material require tool to drop and slows down mining speed if the incorrect tool is used.
```
```suggestion
	 * Make the material require a tool to drop and slows down mining speed if the incorrect tool is used.
```

i509VCB commented on fabric-object-builders-v0/src/main/java/net/fabricmc/fabric/api/block/FabricBlockSettings.java:
```diff
@@ -68,16 +68,27 @@ public static FabricBlockSettings copyOf(Block.Settings settings) {
 
+	/**
+	 * Makes the block breakable by any tool if {@code breakByHand} is set to true.
+	 */
```
I get that docs are nice, but the v0 block settings should be deprecated, so unsure if the effort is needed for it

shedaniel replied:
Just a few neat javadocs, can't go wrong with that tbh

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
 	 *
-	 * @param stack The stack to check on.
-	 * @param user The current user of the tool, or null if there isn't any.
+	 * @param stack The stack mining the block
```
```suggestion
	 * @param stack The item stack being used to mine the block
```

shedaniel replied:
`The item stack used to mine the block` instead?

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
-	 * @param stack The stack to check on.
-	 * @param user The current user of the tool, or null if there isn't any.
+	 * @param stack The stack mining the block
+	 * @param user  The current user of the tool, or null if there isn't any
```
```suggestion
	 * @param user The current user of the tool, or null if there isn't any
```

shedaniel replied:
![image](https://user-images.githubusercontent.com/34910653/80680301-14dcfd80-8af1-11ea-9675-126fe48bf321.png)
we have been adding spaces, **and** not adding spaces at the same time, do we want to bikeshed javadoc spacing

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
+	// nullable on user once we have an official @Nullable annotation in
+	@Deprecated
+	default float getMiningSpeedMultiplier(ItemStack stack, LivingEntity user) {
+		return 1f;
```
Maybe use `1.0F` instead so the code style isn't over the place

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
+	 * @return The mining level of the item. 3 is equal to a diamond pick.
+	 */
+	// nullable on user once we have an official @Nullable annotation in
+	default int getMiningLevel(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
```
What is nullable here?

The user?

If it's the user I would change to this for now:
```suggestion
	default int getMiningLevel(Tag<Item> tag, BlockState state, ItemStack stack, /* @Nullable */ LivingEntity user) {
```

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolHandlers.java:
```diff
@@ -0,0 +1,304 @@
+public class ToolHandlers implements ModInitializer {
+	@Override
+	public void onInitialize() {
+		ToolManagerImpl.general().register(new ModdedToolsModdedBlocksToolHandler());
+		ToolManagerImpl.general().register(new VanillaToolsModdedBlocksToolHandler());
+		ToolManagerImpl.general().register(new VanillaToolsVanillaBlocksToolHandler());
```
Maybe make all these tool handlers their own classes?

Would clean up the nearly 70% of this class that is inner classes. 

Maybe to a `v1.handler` package?

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,273 @@
+	 * Hook for ItemStack.isEffectiveOn and similar methods.
+	 */
+	//TODO: nullable on user once we have an official @Nullable annotation in
+	public static TriState handleIsEffectiveOn(BlockState state, ItemStack stack, LivingEntity user) {
```
specify nullability to the applicable return value or parameters

shedaniel committed:
Resolve some issues

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,273 @@
+		 * @param user  the user involved in breaking the block, null if not applicable.
+		 * @return the result of effectiveness
+		 */
+		default ActionResult isEffectiveOn(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
```
Mark the user param nullable?

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,273 @@
+		 * @param user  the user involved in breaking the block, null if not applicable.
+		 * @return the result of mining speed.
+		 */
+		default TypedActionResult<Float> getMiningSpeedMultiplier(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
```
again like above, mark the user param nullable

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/mixin/tool/attribute/MixinItemStack.java:
```diff
@@ -42,30 +41,12 @@
-			info.setReturnValue(triState.get());
-			info.cancel();
-		}
+		info.setReturnValue(ToolManager.handleIsEffectiveOn(state, (ItemStack) (Object) this, null).get());
```
I horribly dislike the fact that we are using a short circuiting injector to handle this. Other mods which mixin to `isEffectiveOn` will silent fail if they inject after fabric here.

I'd say `@Overwrite` or find a way to keep the vanilla codepath within a possible scope to be executed.

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/mixin/tool/attribute/MixinItemStack.java:
```diff
@@ -42,30 +41,12 @@
-
-			info.setReturnValue(triState.get() ? miningSpeed : 1.0F);
-		}
+		info.setReturnValue(ToolManager.handleBreakingSpeed(state, (ItemStack) (Object) this, null));
```
Similar to above comment about short circuiting injector.

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
-	int getMiningLevel(ItemStack stack, LivingEntity user);
+	// nullable on user once we have an official @Nullable annotation in
+	@Deprecated
+	default int getMiningLevel(ItemStack stack, LivingEntity user) {
```
You need to specify which param is nullable here

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
-	float getMiningSpeedMultiplier(ItemStack stack, LivingEntity user);
+	// nullable on user once we have an official @Nullable annotation in
+	@Deprecated
+	default float getMiningSpeedMultiplier(ItemStack stack, LivingEntity user) {
```
Ditto. Specify nullable thing

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
+	 * @return The mining speed multiplier of the item. 8.0 is equal to a diamond pick.
+	 */
+	// nullable on user once we have an official @Nullable annotation in
+	default float getMiningSpeedMultiplier(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
```
Same again.
Ditto. Specify nullable thing

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
+	 * @param isEffective  whether the tool has been handled
+	 * @return the speed after post processing
+	 */
+	default float postProcessMiningSpeed(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user, float currentSpeed, boolean isEffective) {
```
Same again.
Ditto. Specify nullable thing

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -57,10 +111,10 @@
 	 */
-	//TODO: nullable on user once we have an official @Nullable annotation in
+	// nullable on user once we have an official @Nullable annotation in
 	default Multimap<String, EntityAttributeModifier> getDynamicModifiers(EquipmentSlot slot, ItemStack stack, LivingEntity user) {
```
Same again.
Ditto. Specify nullable thing

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolManager.java:
```diff
@@ -0,0 +1,57 @@
+	 * @param user  the user involved in breaking the block, null if not applicable.
+	 * @return the state of effective
+	 */
+	public static TriState handleIsEffectiveOn(BlockState state, ItemStack stack, LivingEntity user) {
```
okay I know you are getting sick of these nullable mentions, but for maintainability and ease of possible annotations in the future, we need to mark these nullable params

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,273 @@
+			Tag<Item>[] newTags = new Tag[tags.length + 1];
+			int[] newTagLevels = new int[tagLevels.length + 1];
+			System.arraycopy(tags, 0, newTags, 0, tags.length);
+			System.arraycopy(tagLevels, 0, newTagLevels, 0, tagLevels.length);
```
I assume these tags you array copy are either already generated or null checked?

shedaniel replied:
This was from the old tool manager, I renamed the class and now it shows up like this, I _might_ want to move these to lists but I don't have an idea on the performance impact.

shedaniel replied:
well hmm nevermind, it seems like I using arrays might not be a bad idea

Player3324 replied:
Should use Arrays.copyOf though

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,273 @@
+			@Override
+			public ActionResult isEffectiveOn(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
+				for (ToolHandler toolHandler : toolHandlers) {
+					ActionResult effectiveOn = Objects.requireNonNull(toolHandler.isEffectiveOn(tag, state, stack, user));
```
Maybe instead of hard failing here, we just return `FAIL` or `PASS`?

shedaniel replied:
I think you shouldn't be returning null anyway

LemmaEOF replied:
Null-checking catches errors more quickly, so I'd say it's OK.

i509VCB commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/ToolManagerImpl.java:
```diff
@@ -0,0 +1,273 @@
+			@Override
+			public TypedActionResult<Float> getMiningSpeedMultiplier(Tag<Item> tag, BlockState state, ItemStack stack, LivingEntity user) {
+				for (ToolHandler toolHandler : toolHandlers) {
+					TypedActionResult<Float> miningSpeedMultiplier = Objects.requireNonNull(toolHandler.getMiningSpeedMultiplier(tag, state, stack, user));
```
Consider this like the above where ever you call `Objects.requireNonNull`

i509VCB reviewed (changes requested):
Noticed some specific Mixin stuff that is concerning to me at least.

shedaniel committed:
a

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/ToolManager.java:
```diff
@@ -0,0 +1,57 @@
+
+/**
+ * API facing part to register tool handlers and get information about how tools are handled.
+ * Implement {@link DynamicAttributeTool} to change the mining level or speed of your tool.
```
This documentation is outdated

Player3324 commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/api/tool/attribute/v1/DynamicAttributeTool.java:
```diff
@@ -33,22 +36,73 @@
+	}
+
+	/**
+	 * Post process the mining speed.
```
This needs further explanation with examples where it matters and where it doesn't.

Player3324 reviewed (approved):
I think the tool handler could be improved further by scrapping the events and manually selecting the right kind of handler at the call site. It is somewhat confusing how e.g. VanillaToolsModdedBlocksToolHandler gets used with a tool or block that doesn't match its name and first has to reject what it doesn't handle.

Will approve anyway to get the ball rolling.

shedaniel committed:
remove vanilla tools vanilla blocks thing ok

shedaniel committed:
descriptive

shedaniel commented:
I've actually removed the VanillaToolsVanillaBlocksToolHandler in favour of just passing the stack to vanilla to handle it.
See https://github.com/FabricMC/fabric/pull/589/files#diff-ba1cda96d2667d20bf36029d655d9534R44-R45

shedaniel committed:
Get the faster speed instead

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/impl/tool/attribute/handlers/ModdedMiningToolsVanillaBlocksToolHandler.java:
```diff
@@ -0,0 +1,100 @@
+ *
+ * <p>Only applicable to blocks that are vanilla or share the material that is handled by their vanilla tool.</p>
+ */
+public class ModdedMiningToolsVanillaBlocksToolHandler implements ToolManagerImpl.ToolHandler {
```
Maybe have this `ModdedToolsVanillaBlocksToolHandler`?

LemmaEOF commented on fabric-tool-attribute-api-v1/src/main/java/net/fabricmc/fabric/mixin/tool/attribute/MixinItemStack.java:
```diff
@@ -31,40 +31,26 @@
-			info.setReturnValue(triState.get());
-			info.cancel();
+		if (!info.getReturnValueZ()) {
+			info.setReturnValue(ToolManager.handleIsEffectiveOnIgnoresVanilla(state, (ItemStack) (Object) this, null).get());
```
Is it ok to not have this be TriStates anymore?

shedaniel replied:
Yeah I should do that

shedaniel committed:
Rename ModdedToolsVanillaBlocksToolHandler

shedaniel committed:
Remove tri state

shedaniel committed:
Fix compile errors

shedaniel committed:
Merge remote-tracking branch 'fabric/1.15' into patch-1

shedaniel committed:
testmod

shedaniel committed:
fix formatting

shedaniel committed:
fix testmod javadoc

shedaniel commented:
i have added testmods

modmuss50 commented:
I assume this is backwards complaible with existing mods that add tools?

shedaniel commented:
If you mean tools created using the tool attribute api currently, they don't work anyways.
but if you mean the tools before that, it should just pass through because they are basically vanilla tools. mods like materialisation hacks into fabric api itself to get the effect, might be broken, but yeah they are hacks

modmuss50 commented:
Humm ok, it seems in TechReborn I add my tools to the fabric tool tags, and give them a ToolMaterial while extending the vanilla item class (ShovelItem for example).

modmuss50 reviewed (approved):
Im happy with this.

I assume its been tested on client / server. The addition of the test mod is useful.

The netherite items will need adding when merged into 1.16

shedaniel commented:
Oh right, I have completely forgotten to test this on a server, I'd assume that it should work, but I will test it later

LemmaEOF commented:
This is now on last call, pending the server test!

shedaniel committed:
forgot to license format testmod

shedaniel commented:
tested this to be compilable and working on a server

modmuss50 commented:
7 days is up, I will get this in shortly.

modmuss50 commented on fabric-object-builder-api-v1/build.gradle:
```diff
@@ -1,5 +1,5 @@
 archivesBaseName = "fabric-object-builder-api-v1"
-version = getSubprojectVersion(project, "1.0.0")
+version = getSubprojectVersion(project, "1.0.1")
```
I will bump this to 1.1.0

shedaniel replied:
alright

shedaniel committed:
Bump to 1.1.0
